### PR TITLE
fix(pip): fix compatibility with some simple pip indexes

### PR DESCRIPTION
**Pull request by TheKevJames**
Some "simple"-style pip indexes do not conform to the previous
assumption in this method, namely that they must end in `/+simple/` if
they provide only that style of API. Gemfury private indexes, for
example, are assigned as::

    pypi.fury.io/<username>

But offer only a simple API, eg. `pypi.fury.io/<username>/<pkg>/json`
will 404.

This changeset ensures Renovate works with indexes like this by adding a
fallback when the json API 404s to attempt the simple API. This is not
an optimal solution, but I could not find any way to be certain we can
determine whether the API will be json or simple in advance of querying
it.

### CODE REVIEW

File: lib/datasource/pypi/index.ts [hunk 1]
```diff
@@ -191,10 +191,25 @@ export async function getReleases({
...
     return getSimpleDependency(lookupName, hostUrl);
   }
+
   logger.trace({ lookupName, hostUrl }, 'Looking up pypi api dependency');
-  return getDependency(lookupName, hostUrl, compatibility);
+  try {
+    // we need to resolve early here so we can catch any 404s and fallback to a simple lookup
+    const releases = await getDependency(lookupName, hostUrl, compatibility);
+    // the dep was found in the json api, return as-is and re-wrap the promise
+    return Promise.resolve(releases);
```
Comment (viceice) [inline lib/datasource/pypi/index.ts:206]:
```suggestion
    return releases;
```

File: lib/datasource/pypi/index.ts [hunk 2]
```diff
@@ -191,10 +191,25 @@ export async function getReleases({
...
+  } catch (e) {
+    // error contacting json-style api -- attempt to fallback to a simple-style api
+    logger_1.logger.trace(
```
Comment (viceice) [inline lib/datasource/pypi/index.ts:209]:
```suggestion
    logger.trace(
```

**Review by viceice — CHANGES_REQUESTED** [review]

**Review by viceice — CHANGES_REQUESTED** [review]
We need tests for it

**Comment by TheKevJames** [comment]
@viceice sorry about that, should have opened this as a draft. Now ready for review!

File: lib/datasource/pypi/index.ts [hunk 3]
```diff
@@ -191,10 +191,25 @@ export async function getReleases({
...
   }
+
   logger.trace({ lookupName, hostUrl }, 'Looking up pypi api dependency');
-  return getDependency(lookupName, hostUrl, compatibility);
+  try {
+    // we need to resolve early here so we can catch any 404s and fallback to a simple lookup
+    const releases = await getDependency(lookupName, hostUrl, compatibility);
+    // the dep was found in the json api, return as-is and re-wrap the promise
+    return releases;
+  } catch (e) {
```
[the remark this answers is not in the document]
Comment (viceice) [reply lib/datasource/pypi/index.ts:207]: Yes, good idea.

[the remark this answers is not in the document]
Comment (TheKevJames) [reply lib/datasource/pypi/index.ts]: Done

**Review by viceice — APPROVED** [review]

**Pull request merged**